Better string recovery

BAD-AL opened the pull request:
- Adds the ability for the user to add 'known' strings to the dictionary in order to better recover strings.
- Adds support for unmunging of .snd, mus, ffx files

BAD-AL committed:
1. Fixes for mcfg unmunge (chunk_processor.cpp, handle_config)
2. Read strings from dictionary (main.cpp, fnv_hashes.cpp, .h, dictionary.txt)
3. fix compile issue in model_scene.cpp
4. Changed hash 'to_string()' to show hex value instead (fnv_hashes.cpp)

BAD-AL committed:
updated dictionary
Tested unmunge on xbox, pc, psp lvl files.
I did not appear to introduce any new instability.
Still plenty of crashes present from version 1.2.0

BAD-AL committed:
updated readme

BAD-AL committed:
modified code to reference the 'dictionary.txt' in the same directory as the tool (for hashes).
+ added code to also check 'tmp_dict.txt' in the directory the command is execured from (and program directory).

BAD-AL committed:
Added support for sound files (.snd, .mus, .ffx)

BAD-AL committed:
Merge remote-tracking branch 'upstream/master'

SleepKiller commented:
I'm very sorry about the slow response! I didn't have a lot of headspace these past couple days.

There are some changes I'd like to request before merging this to bring it inline with the style of the rest of code base. They're basically entirely my fault for not having contributor guidelines written up however.

Before I do that though there is just a design element for the feature I want to talk about. I notice that this appears to introduce a requirement on `dictionary.txt` being carried alongside the tool. I'm unsure about this requirement as I feel it's surprising for a command line tool to not work because a .txt file was missing from it's directory (a user might not think to copy it around).

For end users I feel like it could be better to maintain the existing builtin list of hashes and provide a command line option to overlay extra strings on top of that. The way to best design that is bit trickier of course. 

One possible way would be to add a `Swbf_fnv_hashes` class that could be constructed in `extract_file` and passed down through the unmunge process the same way `File_saver` currently is.
https://github.com/SleepKiller/swbf-unmunge/blob/d0d2ab5474982adf5899389ccbb0b8dcf3bd9325/src/main.cpp#L29-L48

That classes lookup method could then first try looking in the existing immutable map and if it failed to find the hash in that it could fallback to a map of strings from a file that had been passed through on the command line. And if no file had been passed through on the command line functionality would continue to be as it currently is.

A design such as that would I feel avoid the issue of introducing a subtle dependency on a text file and also allow future extensibility. Such as potentially a feature that collected and built a list of strings from all the .lvls in a mod's addon folder.

All that said I get that that could be quite a bit more work than you'd have wanted to do in opening this PR. So if that is case just say, as there are simpler designs that while less extensible in the future will be perfectly serviceable currently. Such as a half-way between what you have now and how it currently works in 1.2.

BAD-AL commented:
One thing that I intended to add but forgot to was the lvl's filename to hashes (I can do that now).
I played around with ripping through the .lvl file looking for strings, but that effort ended up with minimal helpful results.
Adding the lvl's filename to the hashes is what ended up being most helpful.

I do understand about the concern about the tag-along dictionary and believe that I have an idea on how to accomplish your request elegantly without too much effort.

BAD-AL commented:
One more thing to consider about using an external dictionary as a dependency that the program already is dependent on files in the same directory [fmt.dll, tbb.dll]. So when the user wants to transport the program elsewhere, there is already baggage they must carry around with it.

BAD-AL commented:
Upon further review, we could improve the results by processing the embedded object/odf files first and adding the property values to the hashed strings map. I'm not exactly sure how much re-factoring we'd end up doing for this, but it seems like a non-trivial amount.

Or we could print out the object property values that were previously unknown (with a command line option) so that the user could create their own dictionary with the 'previously unknown' strings and try again (this is certainly easier).

SleepKiller commented:
> One more thing to consider about using an external dictionary as a dependency that the program already is dependent on files in the same directory [fmt.dll, tbb.dll]. So when the user wants to transport the program elsewhere, there is already baggage they must carry around with it.

Yes that is very true. This wasn't the case when I first made swbf-unmunge but I think vcpkg now supports static linking all the dependencies the tool needs. Assuming that is the case I feel it would be a good thing to enable, it'd keep the releases cleaner and make it easier to add other dependencies in the future without fear of cluttering the release folder.

I also do think that people are more likely to think to carry along the .dll files than any .txt files. But I could be wrong and people may just do `Ctrl + A` + `Ctrl + C` in any case though with static linking that would moot.

> Upon further review, we could improve the results by processing the embedded object/odf files first and adding the property values to the hashed strings map. I'm not exactly sure how much re-factoring we'd end up doing for this, but it seems like a non-trivial amount.
> 
> Or we could print out the object property values that were previously unknown (with a command line option) so that the user could create their own dictionary with the 'previously unknown' strings and try again (this is certainly easier).

Something like is what I have in mind with building a list of strings from an .lvl. However I agree that the needed refactoring with attaching such functionality onto the current readers would be problematic. Which is to say nothing of readability concerns after the fact. Bolting on such functionality isn't necessarily going to be clean.

What I sort of have in mind is a preprocessing step that a user could run before extracting any .lvls. Where swbf-unmunge gets a directory and iterates through the .lvls it can find there. This step would build a list of strings found in the .lvls and then that list could be as you say saved out and then passed back into it for use during extraction.

For class definitions I think that we'd probably need to go a step further than simply collecting all the strings form them and actually have some minimal logic to read them properly, only grabbing strings from properties we know to contain them and parsing the strings on properties correctly.

Take this as one random example. In this case it's like `cis_fly_droidfighter_shift_up_low_property` is the part of the string we're interested in so we want to correctly take that part of the string only. I think there are other cases as well they'd need special handling to correctly grab the string.
```
BoostSound              = "cis_fly_droidfighter_shift_up_low_property 0.11 1"
```

We probably especially don't want superfluous strings from properties like `AimerPitchLimits` just in case we get hash collisions, we don't want to accidentally give a user a file named `-90.0 90.0.fx`. Just a contrived made up example of course but the problem is worth at least keeping in mind, however unlikely it is to happen.

I feel like I have something else I was going to add here as well but it's completely slipped my mind atm, lol. If it comes back to me I'll add another comment with it.

BAD-AL commented:
Here is the summary of the requirements I'm envisioning with this discussion :

1. Keep swbf_fnv_hashes.cpp as as it is and give the user the option of using a string dictionary.
1. Add an option to generate a dictionary for a given directory (containing lvl files)
   * Process sub directories too
1. When generating a dictionary, do not save output files (.tga, .msh, ...); only save the dictionary file.
1. If the dictionary file already exists, append to it.
1. When adding found strings and a string contains a space (' ') in it, add the substring(0, index_of_space)
   * Example:
	  When you find:  "cis_fly_droidfighter_shift_up_low_property 0.11 1"
	       only add:  "cis_fly_droidfighter_shift_up_low_property"
1. Try to setup static linking to reduce baggage (fmt.dll tbb.dll).
    * Result: Failure 
		 - Error with 'DirectXTex' package
		 "CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:48 (message):
		  directxtex only supports dynamic crt linkage"

SleepKiller commented:
>     1. Keep swbf_fnv_hashes.cpp as as it is and give the user the option of using a string dictionary.
> 
>     2. Add an option to generate a dictionary for a given directory (containing lvl files)
>        
>        * Process sub directories too
> 
>     3. When generating a dictionary, do not save output files (.tga, .msh, ...); only save the dictionary file.
> 
>     4. If the dictionary file already exists, append to it.
Yup sounds good. Just to clarify as well I feel like the code that collects strings and the extraction code (`extract_***.cpp` files) should be fully separate.


>     5. When adding found strings and a string contains a space (' ') in it, add the substring(0, index_of_space)
>        
>        * Example:
>          When you find:  "cis_fly_droidfighter_shift_up_low_property 0.11 1"
>          only add:  "cis_fly_droidfighter_shift_up_low_property"
Yeah. One thing I will reiterate though is I am concerned about the possibility of collisions from class properties. It is unlikely to ever be a problem but it's not impossible and would be very confusing for someone that ran into it. So while handling those care should be taken to only grab strings from properties of interest. Possibly validating that they don't any [reserved characters](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions) before adding them to the dictionary, the last thing we would want is a file failing to save because it got assigned a name with a `\` in it, in that case we'd want it to gracefully fallback to the current behavior.

>     6. Try to setup static linking to reduce baggage (fmt.dll tbb.dll).
>        
>        * Result: Failure
>          
>          * Error with 'DirectXTex' package
>            "CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:48 (message):
>            directxtex only supports dynamic crt linkage"
Ah you might need to run `git pull` and then `bootstrap-vcpkg.bat` again. I'm pretty sure it should work as I recently switched another project over to static linking DirectXTex.

Anyway I feel like this can be done in two cleanly separate parts. Ignoring static linking that is, which is only really indirectly apart of this and can just be done at "some point" before the next release. First the changes to `swbf_fnv_hashes.cpp` to support loading a secondary set of strings from a file to use. And separately the ability to create a list of strings from a directory of .lvl files. Both things can be worked on and merged separately since they have no direct dependency on each other.

BAD-AL commented:
Based on the shipped ODF file contents I'm going to push back on the requirement on 'reserved characters'.
In some of the shipped odf files /lvls there is stuff like:
```
\\ChunkGeometryName = "yav_bldg_watchtower_hulk"
\\ChunkPhysics = "STATIC"
\\ChunkTrailEffect      = "com_sfx_chunktrail_sm"
\\ChunkSmokeEffect  = "com_sfx_smokeplume"
\\ChunkSmokeNodeName    = "hp_smoke1"
```
These look like they are 'commented out' and I guess they kind of are. But they get put in the object file.
Example (unmunged tur_bldg_tower.odf):
```
tur\odf\tur_bldg_tower.odf:0xdd807455 = "yav_bldg_watchtower_hulk"
tur\odf\tur_bldg_tower.odf:0x8b6e7dc3 = "STATIC"
tur\odf\tur_bldg_tower.odf:0x88e0a44b = "com_sfx_chunktrail_sm"
tur\odf\tur_bldg_tower.odf:0xc74113de = "com_sfx_smokeplume"
tur\odf\tur_bldg_tower.odf:0x7aded410 = "hp_smoke1"
```
If we do not correctly resolve "\\\\ChunkGeometryName", the user will be left wondering what property actually gets the displayed value.

SleepKiller commented:
I was mostly thinking of strings acquired through looking .lvl files. Class property names can't be acquired that way. It does raise interesting point that I had not considered though.

So it looks like the string list does already have some entries for badly commented properties. 
https://github.com/SleepKiller/swbf-unmunge/blob/d0d2ab5474982adf5899389ccbb0b8dcf3bd9325/src/swbf_fnv_hashes.cpp#L46-L53

There are a couple ways to look at this really. The first is that it's a good thing and it let's some .odf files be reconstructed more correctly. Another way is that it's a bad thing and needlessly increases the risk of hash collisions going forward as they don't actually affect the functionality of the resulting object class in anyway.

I'm of two minds here because a sporadic failure from the tool attempting to name a config file `\\ChunkGeometryName` wouldn't be great but it is unlikely to happen. And it could potentially be made a nonissue by simply looking up into the dynamic user supplied map first before trying the constant builtin one. So if we do go that way (or something with an equivalent effect) I'm probably happy to go ahead and add these to the current builtin map if you feel it's worth for aesthetics.
```c++
"\\ChunkGeometryName"_fnvp,
"\\ChunkPhysics"_fnvp,
"\\ChunkTrailEffect"_fnvp,
"\\ChunkSmokeEffect"_fnvp,
"\\ChunkSmokeNodeName"_fnvp,
```

BAD-AL committed:
Synch with swbf-unmunge master (2/10/2021)

BAD-AL committed:
Functionality to read from a dictionary specified from command line

BAD-AL commented:
For now, I'm just committing:
1. The 'dictionary read' stuff.
2. Strings that improve the file unmunge process.

I'd like to create a separate pr for the dictionary build after 'dictionary read' is integrated.

SleepKiller commented:
Sorry about the delay. I'll try to start going through this today as I'm able to.

PrismaticFlower commented on readme.md:
```diff
@@ -36,7 +36,7 @@ Would save the extracted contents of `test.lvl` into a folder named `test`.
 ------------ | -------------
 Object Definitions | Recovered nearly perfectly.
-Config Files (*.fx, *,sky, etc) | Depending on the file type, recovered perfectly or for certain types poorly. In all cases the name of the file is not recovered.
+Config Files (*.fx, *,sky, *.mcfg etc) | Recovered nearly perfectly.
```
```suggestion
With a string dictionary recovered nearly perfectly for most file types or recovered poorly for some file types without one.
```
Until #25 (and automatic dictionary building) is resolved something like this could be the wording to go with for the readme.

PrismaticFlower commented on src/app_options.cpp:
```diff
@@ -289,6 +295,10 @@ Input_platform App_options::input_platform() const noexcept
+std::string App_options::user_string_dict() const noexcept
+{
+   return _user_string_dict;
+}
```
```suggestion
}

```
Very minor style thing, looks like a newline is missing in between the two functions.

PrismaticFlower commented on src/handle_config.cpp:
```diff
@@ -59,22 +59,48 @@ bool is_string_data(Ucfb_reader_strict<"DATA"_mn> data)
 bool is_hash_data(Ucfb_reader_strict<"DATA"_mn> data)
 {
-   const std::array<std::uint32_t, 7> hashes = {
+   bool retVal = false;
```
This is a style issue that's my fault for not having written down a contributor/style guide. swbf-unmunge uses `snake_case` for everything except type names which are capitalized as `Snake_case`.

I also think it could be cleaner keep the current design of simply computing the return value in the return statement. Unless of course there's some good reason for changing it that I just haven't seen. In which case feel free to push back on this.
```suggestion
```

BAD-AL replied:
sorry about the different style.
I do try to use the style of the project I'm in, but camelCase is such a habit that I didn't notice the difference.

PrismaticFlower commented on src/handle_config.cpp:
```diff
@@ -59,22 +59,48 @@ bool is_string_data(Ucfb_reader_strict<"DATA"_mn> data)
+      0x65bfa4e1, /* "Squat" */
+      0x70dc49d0, /* "BodyFall" */
+      0xabd36816,  /* "I3DL2ReverbPreset" */
    };
```
Again a style issue that's my fault for not having guides written up. In swbf-unmunge I made the choice to prefer `std::array` to C style arrays. (`std::array` has a cleaner API and most importantly doesn't decay to a pointer)

More than that I'm not sure why I didn't do this back when I first wrote this file but we can actually make this a lot cleaner to maintain.
```suggestion
   constexpr std::array hashes = {
      "GrassPatch"_fnv,
      "File"_fnv,
      "Sound"_fnv,
      "CollisionSound"_fnv,
      "Path"_fnv,
      "BorderOdf"_fnv,
      "LeafPatch"_fnv,
      "Name"_fnv,
      "Movie"_fnv,
      "Inherit"_fnv,
      "Segment"_fnv,
      "Font"_fnv,
      "Subtitle"_fnv,
      "BUS"_fnv,
      "Stream"_fnv,
      "SoundStream"_fnv,
      "Sample"_fnv,
      "Group"_fnv,
      "Class"_fnv,
      "FootstepLeftWalk"_fnv,
      "FootstepRightWalk"_fnv,
      "FootstepLeftRun"_fnv,
      "FootstepRightRun"_fnv,
      "FootstepLeftStop"_fnv,
      "FootstepRightStop"_fnv,
      "Jump"_fnv,
      "Land"_fnv,
      "Roll"_fnv,
      "Squat"_fnv,
      "BodyFall"_fnv,
      "I3DL2ReverbPreset"_fnv,
   };
```
This uses the _fnv user-defined literal from `swbf_fnv_hashes.hpp` to provide a cleaner look and also uses C++17's [CTAD](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction) to get the same benefits of the C array of having the size deduced.

PrismaticFlower commented on src/handle_config.cpp:
```diff
@@ -59,22 +59,48 @@ bool is_string_data(Ucfb_reader_strict<"DATA"_mn> data)
-   return ((std::find(std::cbegin(hashes), std::cend(hashes), data_hash) !=
-            std::cend(hashes)) &&
-           element_count > 0);
+   
+   if (element_count > 0) {
+      retVal = std::find(std::begin(a), std::end(a), data_hash) != std::end(a);
+   }
+   return retVal;
```
```suggestion

   return ((std::find(std::cbegin(hashes), std::cend(hashes), data_hash) !=
            std::cend(hashes)) &&
           element_count > 0);
```
As said above I feel this is better than having a separate variable just for storing the return value in such a simple function. Also as I said above I'm happy to shift on this if there's some readability issue I'm missing.

BAD-AL replied:
In functions I will often use a variable to hold the return value to make things easier to inspect while debugging. 
If you really can't stand this style, I can change it.

PrismaticFlower replied:
Ah yeah, that makes sense. In terms of keeping it that way beyond debugging it would come down to if we expect this function to be a common stop during debugging in the future or not. If not then when needed while debugging in the future it can always be edited to assign to a variable first before returning.

PrismaticFlower commented on src/handle_config.cpp:
```diff
@@ -290,10 +316,11 @@ void handle_config(Ucfb_reader config, File_saver& file_saver, std::string_view
 {
    const auto name_hash =
       config.read_child_strict<"NAME"_mn>().read_trivial<std::uint32_t>();
+   auto actualFileName = lookup_fnv_hash(name_hash);
```
Same style issue as before. This just needs to be made snake case or even just made `name` since the function is so small it probably won't be ambiguous. But there's also no wrong move in making it longer and more clear.
```suggestion
   const auto name = lookup_fnv_hash(name_hash);
```

PrismaticFlower commented on src/handle_config.cpp:
```diff
@@ -290,10 +316,11 @@ void handle_config(Ucfb_reader config, File_saver& file_saver, std::string_view
 
    if (!buffer.empty()) {
-      file_saver.save_file(buffer, dir, std::to_string(name_hash), file_type);
+      file_saver.save_file(buffer, dir, actualFileName, file_type);
```
This will just need updating to whatever `actualFileName` gets changed to.
```suggestion
      file_saver.save_file(buffer, dir, name, file_type);
```

PrismaticFlower commented on swbf-unmunge.vcxproj:
```diff
@@ -141,6 +141,8 @@
     <Link>
       <SubSystem>Console</SubSystem>
+      <AdditionalDependencies>Shlwapi.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalLibraryDirectories>%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
I don't see anything that needs `Shlwapi.lib` so I'm guessing these are left over from something?
```suggestion
```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -18,6 +19,7 @@
 #include <stdexcept>
 
 #include <Windows.h>
+#include <regex>
```
Another small style thing. I like to keep standard library headers grouped separately from others.
```suggestion
#include <regex>
#include <stdexcept>

#include <Windows.h>
```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -26,6 +28,82 @@ const auto usage = R"(Usage: swbf-unmunge <options>
+                                         "_tdm",
+                                         "_trees",
+                                         "_turrets",
+                                         "_xl"};
```
This can be made const and even constexpr if we change it to store string views instead of strings. We can also change it to an array to save allocating memory at runtime.

On a stylistic note since it's in the global scope of `main.cpp` I feel it should have a more descriptive name to make it clear what it is for.
```suggestion
constexpr std::array common_layer_suffixes{"_1ctf"sv,
                                           "_1flag"sv,
                                           "_Buildings"sv,
                                           "_Buildings"sv,
                                           "_Buildings01"sv,
                                           "_Buildings02"sv,
                                           "_CP-Assult"sv,
                                           "_CP-Conquest"sv,
                                           "_CP-VehicleSpawns"sv,
                                           "_CP-VehicleSpawns"sv,
                                           "_CPs"sv,
                                           "_CommonDesign"sv,
                                           "_CW-Ships"sv,
                                           "_GCW-Ships"sv,
                                           "_Damage"sv,
                                           "_Damage01"sv,
                                           "_Damage02"sv,
                                           "_Death"sv,
                                           "_DeathRegions"sv,
                                           "_Design"sv,
                                           "_Design001"sv,
                                           "_Design002"sv,
                                           "_Design01"sv,
                                           "_Design02"sv,
                                           "_Design1"sv,
                                           "_Design2"sv,
                                           "_Doors"sv,
                                           "_Layer000"sv,
                                           "_Layer001"sv,
                                           "_Layer002"sv,
                                           "_Layer003"sv,
                                           "_Layer004"sv,
                                           "_Light_RG"sv,
                                           "_NewObjective"sv,
                                           "_Objective"sv,
                                           "_Platforms"sv,
                                           "_Props"sv,
                                           "_RainShadow"sv,
                                           "_Roids"sv,
                                           "_Roids01"sv,
                                           "_Roids02"sv,
                                           "_Shadow_RGN"sv,
                                           "_Shadows"sv,
                                           "_Shields"sv,
                                           "_SoundEmmiters"sv,
                                           "_SoundRegions"sv,
                                           "_SoundSpaces"sv,
                                           "_SoundTriggers"sv,
                                           "_Temp"sv,
                                           "_Tree"sv,
                                           "_Trees"sv,
                                           "_Vehicles"sv,
                                           "_animations"sv,
                                           "_campaign"sv,
                                           "_collision"sv,
                                           "_con"sv,
                                           "_conquest"sv,
                                           "_ctf"sv,
                                           "_deathreagen"sv,
                                           "_droids"sv,
                                           "_eli"sv,
                                           "_flags"sv,
                                           "_gunship"sv,
                                           "_hunt"sv,
                                           "_invisocube"sv,
                                           "_light_region"sv,
                                           "_objects01"sv,
                                           "_objects02"sv,
                                           "_reflections"sv,
                                           "_rumble"sv,
                                           "_rumbles"sv,
                                           "_sound"sv,
                                           "_tdm"sv,
                                           "_trees"sv,
                                           "_turrets"sv,
                                           "_xl"sv};

```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -110,6 +189,38 @@ int main(int argc, char* argv[])
+
+         return 0;
+      }
+   }
```
I'm confused by the calls to `std::regex_replace`, it could be obvious but I'm unsure what their intended purpose is?

In any case it feels like we'd need some form of error handling for the `read_dictionary` function, we could either do it locally in `read_dictionary` or surround the call to it with an exception handler (like the one below) and handle it there.
```suggestion

   if (!app_options.user_string_dict().empty()) {
      try {
         read_dictionary(app_options.user_string_dict());
      }
      catch (std::exception& e) {
         synced_cout::print(
            "Error: Exception occured while reading string dictionary.\n   Path: "s,
            app_options.user_string_dict(), '\n', "   Message: "s, e.what(), '\n');
      }
   }
```

BAD-AL replied:
The arguments are processed by app_options which applies 'std::quoted'; here I'm removing the quotes and reducing the ``` '\\' to '\' ```.

I don't really understand why you want to apply quoted stuff in the first place 
Perhaps you can suggest a better solution for that?

PrismaticFlower replied:
Ah I see, are you referring to this use of `std::quoted` or something else? https://github.com/SleepKiller/swbf-unmunge/blob/1d6d0d03dccb0be00b8460bf84378019ac7ec138/src/app_options.cpp#L27
When used on an input stream it should be stripping away quotes (not adding them) so that arguments can be passed in like `-file "C:\Test Folder\file.lvl"`.

I'm still unsure why `\\` is getting replaced with `\` though. If the tool gets passed a path with `\\` in it and the OS rejects it then isn't all the tool needs to do is report the error?

BAD-AL replied:
After further inspection, I resolved the backslash and quote issue in 'App_options.cpp' by instead doing:
```
{"-string_dict"s, [this](Istr& istr) { _user_string_dict = read_file_path(istr); },
```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -110,6 +189,38 @@ int main(int argc, char* argv[])
+      }
+   }
+   else
+      initialize_internal_dictionary();
```
Again a small style thing that's my fault. if statements (and as a result else statements) that don't fit on one line I typically surround with angle brackets.
```suggestion
   else {
      initialize_internal_dictionary();
   }
```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -110,6 +189,38 @@ int main(int argc, char* argv[])
+      for (int j = 0; j < popular_suffixes.size(); j++) {
+         add_hash(tmp + popular_suffixes[j]);
+      }
+   }
```
This loop can probably be made easier to read by using range based for loops and [std::filesystem::path::stem](https://en.cppreference.com/w/cpp/filesystem/path/stem).

```suggestion
   for (const auto& input_file : input_files) {
      const auto name = fs::path{input_file}.stem().string();

      add_hash(name);
      add_hash("mapname.description."s += name);
      add_hash("mapname.name."s += name);

      for (const auto& suffix : common_layer_suffixes) {
         add_hash(std::string{name} += suffix);
      }
   }
```

As a final note both this and the code for loading the string dictionary would probably be better suited to be inside `extract_file` rather than `main`. Or even in their own function for `extract_file` to call.

PrismaticFlower commented on src/swbf_fnv_hashes.hpp:
```diff
@@ -25,4 +25,10 @@ constexpr std::uint32_t operator""_fnv(const char* str, const std::size_t length
+
+void initialize_internal_dictionary();
+
+bool add_hash(std::string str);
```
Given that I (foolishly in retrospect) didn't put anything in a namespaces (like `unmunge::hashes` or something) while first writing swbf-unmunge it might be beneficial to give these more unique names. At the very least something like `read_fnv_dictionary` and `add_fnv_hash` would help make them less generic. 

Though maybe it would pay to put them all (including `lookup_fnv_hash` in namespace or give them all better more unique names like `lookup_swbf_fnv_hash`, `read_swbf_fnv_hash_dictionary`, etc

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
 
 using namespace std::literals;
 
-namespace {
```
The anonymous namespace should be kept to make sure the global variables can't leak out of this file.

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+#include <sstream>
+#include <iostream>
+#include <mutex>
+#include <regex>
```
Same string sstream.
```suggestion
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
-
 #include <cstdint>
+#include <fstream>
+#include <sstream>
```
This doesn't seem to be used? I'm guessing it's left over from something else.
```suggestion
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+#include <iostream>
+#include <mutex>
+#include <regex>
 #include <string>
```
Also include shared_mutex so we can use it instead of `std::mutex` and avoid blocking when multiple threads are just reading from `swbf_hashes`.
```suggestion
#include <shared_mutex>
#include <string>
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
 using namespace std::literals;
 
-namespace {
+std::unordered_map<std::uint32_t, std::string> swbf_hashes;
```
I think the full suggestion here would be too large but I feel the initialization for the builtin hashes for this should be kept here rather than requiring a separate function to be called to do it. In order to ensure that loaded hashes don't replace builtin ones we can simply call [std::unordered_map::emplace](https://en.cppreference.com/w/cpp/container/unordered_map/emplace) which only adds new elements if there isn't already a matching element in the container.

BAD-AL replied:
I was having real trouble with the initialization once the unordered_map was no  longer const.
If you can give an example of proper syntax for this, I'll incorporate it.

PrismaticFlower replied:
That is strange, it works for me just to remove `const` from it. Though thinking about it the change to storing strings rather than string views would require editing the `_fnvp` operator as well, could that be it? This below snippet worked for me locally.
```c++
std::pair<std::uint32_t, std::string> operator""_fnvp(const char* str, std::size_t length)
{
   return {fnv_1a_hash({str, length}), {str, length}};
}

std::unordered_map<std::uint32_t, std::string> swbf_hashes{ // etc...
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
-   "ZoomTurnDivisorMin"_fnvp,
-   "ZOrder"_fnvp};
-}
+std::mutex dictionary_mutex;
```
If we change this to a shared mutex we avoid taking a blocking lock on the mutex when only reading from `swbf_hashes`. Also I like to keep mutexes that are for protecting a specific variable named after that variable. (Again my fault for not having a style guide.)
```suggestion
std::shared_mutex swbf_hashes_mutex;
```

BAD-AL replied:
Please review the changes for mutex handling to ensure I understood your intention.

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+      //synced_cout::print("looked_up:", result->second, "\n");
+      retVal = std::string{result->second};
+   }
+   dictionary_mutex.unlock();
```
We should use [std::shared_lock](https://en.cppreference.com/w/cpp/thread/shared_mutex) rather than directly interacting with the mutex, this way we can be confident we never forgot to unlock it.
```suggestion
   {
      std::shared_lock lock{swbf_hashes_mutex};

      if (const auto result = swbf_hashes.find(hash); result != std::cend(swbf_hashes)) {
         return result->second;
      }
   }
```

BAD-AL replied:
Please review the mutex handling here to ensure I understood your suggestion here

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
 
 #include "swbf_fnv_hashes.hpp"
 #include "synced_cout.hpp"
```
[string_helpers.hpp](https://github.com/SleepKiller/swbf-unmunge/blob/master/src/string_helpers.hpp) has a fast function for converting an integer into a `0x0000` style string. It'll be better to use that rather than `sprintf` later.

```suggestion
#include "string_helpers.hpp"
#include "synced_cout.hpp"
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+                         '\n');
+      retVal = std::string(hex_string);
+   }
+   return retVal;
```
We can reuse `to_hexstring` from string_helpers.hpp to simplify this.
```suggestion
   auto hex_string = to_hexstring(hash);

   synced_cout::print("Warning: Unknown hash looked up.\n"s, "   value: "s, hex_string,
                      '\n');

   return hex_string;
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+}
+
+// Read in a file of strings that can be resolved by the hash lookup
+void read_dictionary(const char* fileName)
```
I prefer not to use C style strings. This should either be `const std::string&` or `const std::filesystem::path&`. (Also there's the naming convention issue as well but I've already mentioned that on a couple other things.)

```suggestion
void read_dictionary(const std::filesystem::path& file_path)
```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+void read_dictionary(const char* fileName)
+{
+   synced_cout::print("Reading dictionary: "s, fileName, '\n');
+   std::ifstream input(fileName);
```
It is never checked if opening the file succeeded. Sadly because IO is one of the C++'s standard library's weaker points we have to do this ourselves by calling [std::ifstream::is_open](https://en.cppreference.com/w/cpp/io/basic_ifstream/is_open).

There is a real style nitpick as well for the most part but I feel this is too dense. Some line breaks help break up the function and make it easier to read (in my opinion, you may very reasonably disagree of course it is subjective after all.)
```suggestion
   synced_cout::print("Reading dictionary: "s, file_path, '\n');

   std::ifstream input(file_path);

   if (!input.is_open()) {
      throw std::runtime_error{"Failed to open file"s};
   }

```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+   for (std::string line; getline(input, line);) {
+      add_hash(line);
+   }
+   input.close();
```
We can skip this call, iostreams thankfully aren't quite isn't this cruel and `std::ifstream` will close the file for us in it's destructor.
```suggestion

```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+   synced_cout::print("Reading dictionary: "s, fileName, '\n');
+   std::ifstream input(fileName);
+   for (std::string line; getline(input, line);) {
+      add_hash(line);
```
It might make more sense to have the check for an empty string here rather than in `add_hash`. Though I'm happy have that argued against.
```suggestion
      if (!line.empty()) add_hash(line);
```

BAD-AL replied:
Hmm...
Since the 'add_hash()' function can be called from other parts of the program, I'd argue that it should be the one to try to ensure that the empty string never enters into the dictionary. It's unlikely that the hash value of the empty string will result in conflicts, but it is a simple check that has little impact on performance.

PrismaticFlower replied:
If you feel that's better then by all means keep the empty check in add hash function. Long term I think we should transition towards using `Swbf_fnv_hashes` class that we can pass throughout the extraction as a `const Swbf_fnv_hashes&` to avoid having mutable global state so it being called from other parts of the program should be moot eventually. That can come later though and for now this will be fine.

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,2834 +1,3935 @@
+      }
+      dictionary_mutex.unlock();
+   }
+   return retVal;
```
Using [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard) (which is like `std::shared_lock` except it blocks until it has sole ownership of the mutex, which we need in order to safely write to `swbf_hashes`)  and [structured bindings](https://en.cppreference.com/w/cpp/language/structured_binding) ([this](https://www.youtube.com/watch?v=aBZlbb9sE-g) is probably a better place to learn them if you're unfamiliar and curious) we can vastly simplify this function, making it both safer and easier to read.


```suggestion
   std::lock_guard lock{swbf_hashes_mutex};

   const auto [element_iterator, inserted] = swbf_hashes.emplace(fnv_1a_hash(str), str);

   return inserted;
```

BAD-AL replied:
Please review mutex handling to ensure I understood your suggestion here.

PrismaticFlower reviewed (commented):
This might be a bit long sorry. Plenty of it is just due to me not having written up proper guides beforehand. I should try and make a better effort to write up some proper guides and the info on the style I've followed while writing swbf-unmunge for anyone in the future to read.

BAD-AL committed:
Changes based on code review requests.

BAD-AL committed:
more code suggestion changes from review.

BAD-AL commented:
I pushed a couple commits to incorporate your suggestions.
I resolved most of the comments with your suggestions. 
Just a bit more conversation left on this one though.

BAD-AL committed:
more changes based on code review conversation.

BAD-AL commented:
Pushed in another commit.
Please review the mutex handling in 'swbf_fnv_hashes.cpp' to ensure that I properly addressed your concerns.

PrismaticFlower commented on src/handle_config.cpp:
```diff
@@ -59,22 +59,47 @@ bool is_string_data(Ucfb_reader_strict<"DATA"_mn> data)
+      return std::find(std::begin(hashes), std::end(hashes), data_hash) != std::end(hashes);
+   }
+   return false;
 }
```
When I ran clang-format over this file it created some formatting differences.
```suggestion
      "I3DL2ReverbPreset"_fnv,
   };

   const auto data_hash = data.read_trivial<std::uint32_t>();
   const auto element_count = data.read_trivial_unaligned<std::uint8_t>();

   if (element_count > 0) {
      return std::find(std::begin(hashes), std::end(hashes), data_hash) !=
             std::end(hashes);
   }

   return false;
}
```

PrismaticFlower commented on readme.md:
```diff
@@ -36,7 +36,7 @@ Would save the extracted contents of `test.lvl` into a folder named `test`.
 ------------ | -------------
 Object Definitions | Recovered nearly perfectly.
-Config Files (*.fx, *,sky, etc) | Depending on the file type, recovered perfectly or for certain types poorly. In all cases the name of the file is not recovered.
+Config Files (*.fx, *,sky, *.mcfg etc) | With a string dictionary recovered nearly perfectly for most file types or recovered poorly for some file types without one
```
Extremely minor thing here, a `.` is missing from the end of the sentence.
```suggestion
Config Files (*.fx, *,sky, *.mcfg etc) | With a string dictionary recovered nearly perfectly for most file types or recovered poorly for some file types without one.
```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -19,13 +20,89 @@
+                                         "_tdm",
+                                         "_trees",
+                                         "_turrets",
+                                         "_xl"};
```
Another clang-format difference (plus an added new line between the array and function).
```suggestion
                                           "_1flag",
                                           "_Buildings",
                                           "_Buildings01",
                                           "_Buildings02",
                                           "_CP-Assult",
                                           "_CP-Conquest",
                                           "_CP-VehicleSpawns",
                                           "_CP-VehicleSpawns",
                                           "_CPs",
                                           "_CommonDesign",
                                           "_CW-Ships",
                                           "_GCW-Ships",
                                           "_Damage",
                                           "_Damage01",
                                           "_Damage02",
                                           "_Death",
                                           "_DeathRegions",
                                           "_Design",
                                           "_Design001",
                                           "_Design002",
                                           "_Design01",
                                           "_Design02",
                                           "_Design1",
                                           "_Design2",
                                           "_Doors",
                                           "_Layer000",
                                           "_Layer001",
                                           "_Layer002",
                                           "_Layer003",
                                           "_Layer004",
                                           "_Light_RG",
                                           "_NewObjective",
                                           "_Objective",
                                           "_Platforms",
                                           "_Props",
                                           "_RainShadow",
                                           "_Roids",
                                           "_Roids01",
                                           "_Roids02",
                                           "_Shadow_RGN",
                                           "_Shadows",
                                           "_Shields",
                                           "_SoundEmmiters",
                                           "_SoundRegions",
                                           "_SoundSpaces",
                                           "_SoundTriggers",
                                           "_Temp",
                                           "_Tree",
                                           "_Trees",
                                           "_Vehicles",
                                           "_animations",
                                           "_campaign",
                                           "_collision",
                                           "_con",
                                           "_conquest",
                                           "_ctf",
                                           "_deathreagen",
                                           "_droids",
                                           "_eli",
                                           "_flags",
                                           "_gunship",
                                           "_hunt",
                                           "_invisocube",
                                           "_light_region",
                                           "_objects01",
                                           "_objects02",
                                           "_reflections",
                                           "_rumble",
                                           "_rumbles",
                                           "_sound",
                                           "_tdm",
                                           "_trees",
                                           "_turrets",
                                           "_xl"};


```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -111,6 +189,40 @@ int main(int argc, char* argv[])
+      }
+   }
+   
+   
```
Another clang-format difference on my end.
```suggestion

```

PrismaticFlower commented on src/chunk_processor.cpp:
```diff
@@ -213,8 +213,24 @@ const auto chunk_processors = Chunk_processor_map{
+        handle_config(args.chunk, args.file_saver, ".ffx"sv, "config"sv, true);
+     }}},
+
 
```
Another little clang-format difference on my end.
```suggestion
```

PrismaticFlower commented on src/main.cpp:
```diff
@@ -111,6 +189,40 @@ int main(int argc, char* argv[])
       return 0;
    }
 
+   if (app_options.user_string_dict().length() > 0) {
+
+      if (fs::exists(app_options.user_string_dict())) {
+         try {
+            read_fnv_dictionary(app_options.user_string_dict());
+         }
+         catch (std::exception& e) {
+            synced_cout::print(
+               "Error: Exception occured while reading string dictionary.\n   Path: "s,
+               app_options.user_string_dict(), '\n', "   Message: "s, e.what(), '\n');
+         }
+      }
+      else {
+         std::cout << "Error: file '"s << app_options.user_string_dict()
+                   << "' does not exist\n";
+
+         return 0;
+      }
+   }
+   
+   
+   // add the input file names and filename + 'popular suffixes' to the hashes
+   for (const auto& input_file : input_files) {
+      const auto name = fs::path{input_file}.stem().string();
+
+      add_fnv_hash(name);
+      add_fnv_hash("mapname.description."s += name);
+      add_fnv_hash("mapname.name."s += name);
+
+      for (const auto& suffix : common_layer_suffixes) {
+         add_fnv_hash(std::string{name} += suffix);
+      }
+   }
+
```
If there isn't a reason for this to be in main (I can't see one, the other modes don't do hash lookups) it can probably be moved into the top of `extract_file`, Or probably into it's own function that's called from `extract_file`.

BAD-AL replied:
My feeling here is that we'd want to populate the string/hash lookup map before processing any files to give more opportunity for the hashes to be found in the process of un-munging related .lvls.

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -1,31 +1,34 @@
 #include <unordered_map>
 #include <utility>
 
 using namespace std::literals;
```
Small style issue, there should be newline between this and the namespace below.
```suggestion
using namespace std::literals;

```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -2819,16 +3868,59 @@ const std::unordered_map<std::uint32_t, std::string_view> swbf_hashes{
+   std::string ret_val;
+
+   std::lock_guard lock{swbf_hashes_mutex};
    const auto result = swbf_hashes.find(hash);
 
-   if (result != std::cend(swbf_hashes)) return std::string{result->second};
+   if (result != std::cend(swbf_hashes)) {
+      // this can be interesting when building a dict
+      // synced_cout::print("looked_up:", result->second, "\n");
+      ret_val = std::string{result->second};
+   }
```
> Please review mutex handling to ensure I understood your suggestion here.
Yup, this is safe. The mutex will be locked from this line until the scope exits, protecting access to `swbf_hashes`. But we can still improve it in two ways.


The first is quite simple since we only need to hold the lock while we access `swbf_hashes` and not for the rest of the function if we put that access in it's own scope (or even function) we limit the time we're holding the lock.
```suggestion
   {
      std::lock_guard lock{swbf_hashes_mutex};

      if (const auto result = swbf_hashes.find(hash); result != std::cend(swbf_hashes)) {
         ret_val = result->second;
      }
   }
```

The second is because we're only [reading](https://en.cppreference.com/w/cpp/container#Thread_safety) from `swbf_hashes` we don't actually need exclusive ownership, we just need to know that no other thread is going to write to it while we're reading from it. This is where the change to `std::shared_mutex` I talked about before could come into play, because with that we can then use [std::shared_lock](https://en.cppreference.com/w/cpp/thread/shared_lock) which will allow us to have multiple threads reading from it without blocking on each other. But if for whatever reason a thread has to write to it (which it admittedly shouldn't be doing in the middle of extracting a file but it is still mutable global state so we should maintain thread safety for it) the `std::lock_guard` in `add_fnv_hash` will take exclusive ownership of the mutex and stop any other thread accessing `swbf_hashes` until it's finished writing to it.
```suggestion
   {
      std::shared_lock lock{swbf_hashes_mutex};

      if (const auto result = swbf_hashes.find(hash); result != std::cend(swbf_hashes)) {
         ret_val = result->second;
      }
   }
```

Another small note is that because `swbf_hashes` is no longer storing string views the explicit construction to `std::string` can be dropped as well.

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -2819,16 +3868,59 @@ const std::unordered_map<std::uint32_t, std::string_view> swbf_hashes{
+      synced_cout::print("Warning: Unknown hash looked up.\n"s, "   value: "s, hex_string,
+                         '\n');
+      ret_val = hex_string;
+   }
```
Another very minor style thing, missing a newline between here and the return statement.
```suggestion
   }

```

PrismaticFlower commented on src/swbf_fnv_hashes.cpp:
```diff
@@ -2819,16 +3868,59 @@ const std::unordered_map<std::uint32_t, std::string_view> swbf_hashes{
+   }
+
+   if (ret_val.empty()) {
+      std::string hex_string = to_hexstring(hash);
+
+      synced_cout::print("Warning: Unknown hash looked up.\n"s, "   value: "s, hex_string,
+                         '\n');
+      ret_val = hex_string;
```
We can save a copy here by just putting the hex string directly into `ret_val` then printing from that. Admittedly the string returned by `to_hexstring` is likely to always fit in the small string buffer so eliminating the copy doesn't save much but it also makes the code more concise without really harming readability (unless you think it does?) so may as well.
```suggestion
      ret_val = to_hexstring(hash);

      synced_cout::print("Warning: Unknown hash looked up.\n"s, "   value: "s, ret_val,
                         '\n');
```

PrismaticFlower reviewed (commented):
It's looking good, most of my suggestions here are just formatting things to keep it inline with what clang-format will do. (Or at least does for me.) There are a couple other notes but nothing too big. All in all it's looking pretty good.

BAD-AL committed:
Update readme.md

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
Update src/main.cpp

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
Update src/swbf_fnv_hashes.cpp

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
Update src/swbf_fnv_hashes.cpp

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
Update src/swbf_fnv_hashes.cpp

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
Update src/chunk_processor.cpp

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
Update src/main.cpp

Co-authored-by: SleepKiller <SleepKiller@users.noreply.github.com>

BAD-AL committed:
a couple more code-review based changes.

BAD-AL commented:
Committed a few more changes; I think all the concerns have been addressed.
Please review to ensure.

Commit:
extremely small quick fixes before merging

- clang-format difference in main.cpp locally
- `#include <mutex>` in swbf_fnv_hashes.cpp
- defer extra readme.md section in anticipation of a proper contributing.md being written

SleepKiller commented:
I'm just about to hit merge on this but a quick note is I have removed the extra section in anticipation of writing a proper CONTRIBUTING.md. 

It's pretty much all looking good though and seems to work great. Thanks!
